utils/filename.go

```diff
@@ -0,0 +1,59 @@
+package utils
+
+import (
+	"regexp"
+	"strings"
+)
+
+const lowerHex = "0123456789abcdef"
+
+var allowedChars = regexp.MustCompile(`[^0-9a-z!@#$&()\[\]{},.;~\-_=+]`)
+
+// These are reserved words and symbols which cannot be a file name, with or without a file extension
+// Mainly special Windows devices
+var reservedWords = regexp.MustCompile(`^aux|^com[0-9¹²³]|^con|^lpt[0-9¹²³]|^nul|^prn`)
+
+// Symbols that cannot (or should not) make up the entirety of a filename
+var illegalSolo = regexp.MustCompile(`^\.+$`)
+
+// Symbols that cannot (or should not) begin a file name
```

Comment:
Starting with a file name with a dash is actually valid in all OSes, as far as I could tell. But, it seems to be very common that shells would handle that poorly and interpret the file name as a flag instead, which raises the possibility of shell injection. It seemed safer this way.

Reply:
the goal of the name mangling here is not to protect the user from bad file names but to provide a consistent experience to people who vendor files across OSs. Files in the vendor or cache directories should be rarely accessed by end users and more often they will just be noticed because they will appear in code reviews